faiss_vector_cache.go

```diff
@@ -18,92 +18,130 @@
 package zap
 
 import (
+	"encoding/binary"
 	"sync"
 	"sync/atomic"
 	"time"
 
+	"github.com/RoaringBitmap/roaring"
 	faiss "github.com/blevesearch/go-faiss"
 )
 
-type vectorIndexCache struct {
-	closeCh chan struct{}
-	m       sync.RWMutex
-	cache   map[uint16]*cacheEntry
-}
-
-type ewma struct {
-	alpha float64
-	avg   float64
-	// every hit to the cache entry is recorded as part of a sample
-	// which will be used to calculate the average in the next cycle of average
-	// computation (which is average traffic for the field till now). this is
-	// used to track the per second hits to the cache entries.
-	sample uint64
-}
-
-type cacheEntry struct {
-	tracker *ewma
-	// this is used to track the live references to the cache entry,
-	// such that while we do a cleanup() and we see that the avg is below a
-	// threshold we close/cleanup only if the live refs to the cache entry is 0.
-	refs  int64
-	index *faiss.IndexImpl
-}
-
 func newVectorIndexCache() *vectorIndexCache {
 	return &vectorIndexCache{
 		cache:   make(map[uint16]*cacheEntry),
 		closeCh: make(chan struct{}),
 	}
 }
 
+type vectorIndexCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
 func (vc *vectorIndexCache) Clear() {
 	vc.m.Lock()
 	close(vc.closeCh)
 
 	// forcing a close on all indexes to avoid memory leaks.
 	for _, entry := range vc.cache {
-		entry.closeIndex()
+		entry.close()
 	}
 	vc.cache = nil
 	vc.m.Unlock()
 }
 
-func (vc *vectorIndexCache) loadVectorIndex(fieldID uint16,
-	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
-	cachedIndex, present := vc.isIndexCached(fieldID)
-	if present {
-		vecIndex = cachedIndex
-		vc.incHit(fieldID)
+func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte, except *roaring.Bitmap) (
+	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, err error) {
+	entry := vc.fetch(fieldID)
+	if entry != nil {
+		index, vecDocIDMap = entry.load()
+		if except != nil && !except.IsEmpty() {
+			for vecID, docID := range vecDocIDMap {
```

Comment:
@Likith101, a small doubt here (for future reference as well). I remember you were caching the vecIDsToExclude as well. I realize that it's removed in the latest patch because it can have compute dent while maintaining it in cases the bitmapts are updated etc. Were you able to gauge how much overhead traversing the map and compute the vecIDsToExclude had? My doubt over here could pointing out something very niche but I suppose it's worth keeping in mind if it does have an overhead.

Reply:
The overhead would only show up when there are constant mutations in the index altering the except bitmap. I was not able to determine how much difference both the approaches made using my testing.